### PR TITLE
[Core] Support labels for ray.remote

### LINKED ISSUE #48476: [Core] Update `get_local_ongoing_lineage_reconstruction_tasks` to return task metadata
### Description

We need a way to match tasks returned by `get_local_ongoing_lineage_reconstruction_tasks` to specific Ray Data operators. Currently, we can't do this because:
1. The function doesn't return task metadata
2. We can't specify metadata for actor tasks, and using the information returned by the function, we can't access the metadata for the actor task's actor.

### Use case

This is necessary for Ray Data to be able to properly account for lineage reconstruction tasks in its resource management

1. Normal Tasks need to be able to be attributed to particular Operator that spun them up (hence the need for `_metadata`)
2. Actor Tasks need to be able to be attributed to their Actor

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?

Allows users to set labels for actors and tasks (we already support node labels): `ray.remote(_labels={"key1": "value1"})` We start with private option but it should be a generally useful thing to be public in the future.

For now it will be used by Data and `get_local_ongoing_lineage_reconstruction_tasks()` API.

## Related issue number
Closes #48476

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_raylet.pyx [hunk 1]
```diff
@@ -737,11 +737,26 @@ cdef class Language:
...
+
+    if label_dict is None:
+        return 0
+
+    for key, value in label_dict.items():
+        if not isinstance(key, str):
+            raise ValueError(f"Label key must be string, but got {type(key)}")
+        if not isinstance(value, str):
+            raise ValueError(f"Label value must be string, but got {type(value)}")
+        label_map[0][key.encode("ascii")] = value.encode("ascii")
```
Comment (raulchen) [inline python/ray/_raylet.pyx:752]: nit, probably should use utf-8, or add a try-catch for non-ascii chars.

File: src/ray/core_worker/common.h [hunk 2]
```diff
@@ -285,11 +291,11 @@ template <>
 struct hash<ray::rpc::LineageReconstructionTask> {
   size_t operator()(const ray::rpc::LineageReconstructionTask &task) const {
     size_t hash = std::hash<std::string>()(task.name());
-    for (const auto &resource : task.resources()) {
-      hash ^= std::hash<std::string>()(resource.first);
-      hash ^= std::hash<double>()(resource.second);
-    }
     hash ^= std::hash<ray::rpc::TaskStatus>()(task.status());
+    for (const auto &label : task.labels()) {
```
Comment (raulchen) [inline src/ray/core_worker/common.h:295]: nit, should sort the label keys to produce consistent hash result, or just use an ordered_map.

File: python/ray/tests/test_advanced.py [hunk 3]
```diff
@@ -158,16 +181,36 @@ def task(counter):
...
         lineage_reconstruction_tasks = get_local_ongoing_lineage_reconstruction_tasks()
-        return (
-            len(lineage_reconstruction_tasks) == 1
-            and lineage_reconstruction_tasks[0][0].name == "task"
-            and lineage_reconstruction_tasks[0][0].resources == {"worker": 1.0}
+        assert len(lineage_reconstruction_tasks) == 2
+        assert {
+            lineage_reconstruction_tasks[0][0].name,
+            lineage_reconstruction_tasks[1][0].name,
+        } == {"task", "Actor.run"}
```
Comment (raulchen) [inline python/ray/tests/test_advanced.py:188]: Is there a guarantee for the order?

**Review by raulchen — APPROVED** [review]
LGTM except for a few small comments

File: python/ray/tests/test_advanced.py (hunk 3 above)
Comment (jjyao) [reply python/ray/tests/test_advanced.py:188]: No guarantee for the order.

File: src/ray/core_worker/common.h (hunk 2 above)
Comment (jjyao) [reply src/ray/core_worker/common.h:295]: Actually we are using `^` so the order doesn't matter.

File: python/ray/tests/test_advanced.py (hunk 3 above)
Comment (raulchen) [reply python/ray/tests/test_advanced.py:188]: then we should sort the results here?

Comment (raulchen) [reply python/ray/tests/test_advanced.py]: yes, otherwise the test can be flaky

**Pull request merged**